File: src/main/java/ninja/hudy/infosmog/reader/HttpReader.java
package ninja.hudy.infosmog.reader;

import ninja.hudy.infosmog.exception.HttpCommunicationException;
import org.hibernate.validator.constraints.NotBlank;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.*;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestTemplate;

import javax.validation.constraints.NotNull;

public abstract class HttpReader extends Reader {
    @Autowired
    private RestTemplate restTemplate;

    protected String sendPost(@NotBlank String address, @NotNull MultiValueMap<String, String> params) throws HttpCommunicationException {
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);

        HttpEntity<MultiValueMap<String, String>> request = new HttpEntity<>(params, headers);

        ResponseEntity<String> response = restTemplate.postForEntity(address, request, String.class);

        if(!HttpStatus.OK.equals(response.getStatusCode())) {
            throw new HttpCommunicationException("Send http request to " + address + " failed with code: " + response.getStatusCodeValue());
        }

        return response.getBody();
    }

    protected String sendGet(@NotBlank String address) throws HttpCommunicationException {
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);

        ResponseEntity<String> response = restTemplate.getForEntity(address, String.class);

        if(!HttpStatus.OK.equals(response.getStatusCode())) {
            throw new HttpCommunicationException("Send http request to " + address + " failed with code: " + response.getStatusCodeValue());
        }

        return response.getBody();
    }
}

File: src/main/kotlin/pl/org/smok/infosmog/reader/HttpReader.kt
package pl.org.smok.infosmog.reader

import pl.org.smok.infosmog.exception.HttpCommunicationException
import org.hibernate.validator.constraints.NotBlank
import org.springframework.http.*
import org.springframework.util.MultiValueMap
import org.springframework.web.client.RestTemplate
import pl.org.smok.infosmog.repository.CityRepository
import pl.org.smok.infosmog.repository.MeasurementRepository
import pl.org.smok.infosmog.repository.MeasurementTypeRepository
import pl.org.smok.infosmog.repository.StationRepository

import javax.validation.constraints.NotNull

abstract class HttpReader(val restTemplate: RestTemplate,
                          measurementTypeRepository: MeasurementTypeRepository,
                          measurementRepository: MeasurementRepository,
                          stationRepository: StationRepository,
                          cityRepository: CityRepository) :
        Reader( measurementTypeRepository,
                measurementRepository,
                stationRepository,
                cityRepository) {

    @Throws(HttpCommunicationException::class)
    protected fun sendPost(@NotBlank address: String, @NotNull params: MultiValueMap<String, String>): String {
        val headers = HttpHeaders()
        headers.contentType = MediaType.APPLICATION_FORM_URLENCODED

        val request = HttpEntity(params, headers)

        val response = restTemplate.postForEntity(address, request, String::class.java)

        if (HttpStatus.OK != response.statusCode) {
            throw HttpCommunicationException("Send http request to " + address + " failed with code: " + response.statusCodeValue)
        }

        return response.body
    }

    @Throws(HttpCommunicationException::class)
    protected fun sendGet(@NotBlank address: String): String {
        val headers = HttpHeaders()
        headers.contentType = MediaType.APPLICATION_FORM_URLENCODED

        val response = restTemplate.getForEntity(address, String::class.java)

        if (HttpStatus.OK != response.statusCode) {
            throw HttpCommunicationException("Send http request to " + address + " failed with code: " + response.statusCodeValue)
        }

        return response.body
    }
}
